File: src-tauri/src/domain/mapping.rs
use std::collections::HashSet;

use super::{DbKind, MappingError, NamedRow, Row, RowSet, ValidationError, Value};

pub fn validate_source_statement(sql: &str) -> Result<(), ValidationError> {
    validate_statement(sql, true)
}

pub fn validate_target_statement(kind: DbKind, sql: &str) -> Result<(), ValidationError> {
    match kind {
        DbKind::Oracle => validate_statement(sql, false),
    }
}

pub fn validate_row_set_columns(row_set: &RowSet, binds: &[String]) -> Result<(), MappingError> {
    let metadata_row = Row::from_columns(
        row_set
            .columns
            .iter()
            .cloned()
            .map(|column| (column, Value::Null))
            .collect(),
    );
    map_row(&metadata_row, binds).map(|_| ())
}

pub fn extract_named_binds(sql: &str) -> Result<Vec<String>, MappingError> {
    let mut binds = Vec::new();
    let mut seen = HashSet::new();

    for bind in extract_named_bind_occurrences(sql)? {
        if seen.insert(bind.to_ascii_uppercase()) {
            binds.push(bind);
        }
    }

    Ok(binds)
}

/// Returns every named bind in source order, retaining repeated placeholders.
///
/// [`extract_named_binds`] remains the deduplicated mapping contract. Drivers
/// that bind positionally use this occurrence-preserving variant instead.
pub fn extract_named_bind_occurrences(sql: &str) -> Result<Vec<String>, MappingError> {
    let bytes = sql.as_bytes();
    let mut binds = Vec::new();
    let mut index = 0;

    while index < bytes.len() {
        match bytes[index] {
            b'\'' | b'"' => index = skip_quoted(bytes, index),
            b'-' if bytes.get(index + 1) == Some(&b'-') => {
                index = skip_line_comment(bytes, index + 2)
            }
            b'/' if bytes.get(index + 1) == Some(&b'*') => {
                index = skip_block_comment(bytes, index + 2)
            }
            b':' => {
                let start = index + 1;
                if bytes.get(start).is_some_and(|byte| is_bind_start(*byte)) {
                    let mut end = start + 1;
                    while bytes.get(end).is_some_and(|byte| is_bind_continue(*byte)) {
                        end += 1;
                    }

                    let bind = &sql[start..end];
                    binds.push(bind.into());
                    index = end;
                } else if bytes.get(start).is_some_and(u8::is_ascii_digit) {
                    let mut end = start + 1;
                    while bytes.get(end).is_some_and(u8::is_ascii_digit) {
                        end += 1;
                    }
                    return Err(MappingError::NumericBind {
                        parameter: sql[start..end].into(),
                    });
                } else {
                    index += 1;
                }
            }
            _ => index += 1,
        }
    }

    Ok(binds)
}

fn validate_statement(sql: &str, source: bool) -> Result<(), ValidationError> {
    let lexical = lexical_sql(sql)?;
    let tokens = lexical
        .split(|character: char| !character.is_ascii_alphanumeric() && character != '_')
        .filter(|token| !token.is_empty())
        .map(|token| token.to_ascii_uppercase())
        .collect::<Vec<_>>();

    let has_expected_first_keyword = if source {
        matches!(tokens.first().map(String::as_str), Some("SELECT" | "WITH"))
    } else {
        matches!(tokens.first().map(String::as_str), Some("INSERT" | "UPDATE" | "MERGE"))
    };
    if !has_expected_first_keyword {
        return Err(ValidationError::new(if source {
            "source SQL must begin with SELECT or WITH"
        } else {
            "Oracle target SQL must begin with INSERT, UPDATE, or MERGE"
        }));
    }

    if tokens.iter().any(|token| {
        matches!(
            token.as_str(),
            "CREATE"
                | "ALTER"
                | "DROP"
                | "TRUNCATE"
                | "RENAME"
                | "GRANT"
                | "REVOKE"
                | "COMMIT"
                | "ROLLBACK"
                | "SAVEPOINT"
                | "BEGIN"
                | "DECLARE"
                | "EXECUTE"
        )
    }) || tokens.windows(2).any(|pair| pair == ["SET", "TRANSACTION"])
    {
        return Err(ValidationError::new(
            "SQL contains a prohibited administrative, transaction, or PL/SQL statement",
        ));
    }

    if source
        && tokens.iter().any(|token| {
            matches!(
                token.as_str(),
                "INSERT" | "UPDATE" | "DELETE" | "MERGE" | "LOCK"
            )
        })
    {
        return Err(ValidationError::new("source SQL must be read-only"));
    }

    if contains_numeric_bind(&lexical) {
        return Err(ValidationError::new(
            "numeric bind placeholders are not supported",
        ));
    }

    Ok(())
}

fn lexical_sql(sql: &str) -> Result<String, ValidationError> {
    let bytes = sql.as_bytes();
    let mut lexical = vec![b' '; bytes.len()];
    let mut index = 0;

    while index < bytes.len() {
        match bytes[index] {
            b'\'' | b'"' => {
                let next = skip_quoted(bytes, index);
                if next == bytes.len() && bytes.last() != Some(&bytes[index]) {
                    return Err(ValidationError::new(
                        "SQL contains an unterminated quoted literal",
                    ));
                }
                index = next;
            }
            b'-' if bytes.get(index + 1) == Some(&b'-') => {
                index = skip_line_comment(bytes, index + 2)
            }
            b'/' if bytes.get(index + 1) == Some(&b'*') => {
                let next = skip_block_comment(bytes, index + 2);
                if next == bytes.len() && !bytes.ends_with(b"*/") {
                    return Err(ValidationError::new(
                        "SQL contains an unterminated block comment",
                    ));
                }
                index = next;
            }
            byte => {
                lexical[index] = byte;
                index += 1;
            }
        }
    }

    let lexical = String::from_utf8(lexical).expect("SQL bytes remain valid UTF-8 after masking");
    if let Some(semicolon) = lexical.find(';') {
        if lexical[semicolon + 1..]
            .chars()
            .any(|character| !character.is_whitespace())
        {
            return Err(ValidationError::new(
                "multiple SQL statements are not supported",
            ));
        }
    }
    Ok(lexical)
}

fn contains_numeric_bind(sql: &str) -> bool {
    sql.as_bytes()
        .windows(2)
        .any(|window| window[0] == b':' && window[1].is_ascii_digit())
}

pub fn map_row(row: &Row, binds: &[String]) -> Result<NamedRow, MappingError> {
    let normalized = row.normalized_index()?;

    binds
        .iter()
        .map(|bind| {
            normalized
                .get(&bind.to_ascii_uppercase())
                .cloned()
                .map(|value| (bind.clone(), value))
                .ok_or_else(|| MappingError::MissingSourceColumn {
                    parameter: bind.clone(),
                })
        })
        .collect()
}

fn is_bind_start(byte: u8) -> bool {
    byte.is_ascii_alphabetic() || byte == b'_'
}

fn is_bind_continue(byte: u8) -> bool {
    is_bind_start(byte) || byte.is_ascii_digit() || matches!(byte, b'$' | b'#')
}

fn skip_quoted(bytes: &[u8], mut index: usize) -> usize {
    let quote = bytes[index];
    index += 1;

    while index < bytes.len() {
        if bytes[index] == quote {
            if bytes.get(index + 1) == Some(&quote) {
                index += 2;
            } else {
                return index + 1;
            }
        } else {
            index += 1;
        }
    }

    index
}

fn skip_line_comment(bytes: &[u8], mut index: usize) -> usize {
    while index < bytes.len() && bytes[index] != b'\n' {
        index += 1;
    }
    index
}

fn skip_block_comment(bytes: &[u8], mut index: usize) -> usize {
    while index < bytes.len() {
        if bytes[index] == b'*' && bytes.get(index + 1) == Some(&b'/') {
            return index + 2;
        }
        index += 1;
    }
    index
}

File: electron/domain/mapping.ts
import type { DomainValue, NamedRow, Row } from "./models";

export type SourceRow = Row;
export type { NamedRow } from "./models";

export class MappingError extends Error {
  readonly kind: "missing_source_column" | "duplicate_source_column" | "numeric_bind";
  readonly parameter: string;

  private constructor(
    kind: MappingError["kind"],
    parameter: string,
    message: string,
  ) {
    super(message);
    this.name = "MappingError";
    this.kind = kind;
    this.parameter = parameter;
  }

  static missingSourceColumn(parameter: string): MappingError {
    return new MappingError(
      "missing_source_column",
      parameter,
      `missing source column: ${parameter}`,
    );
  }

  static duplicateSourceColumn(column: string): MappingError {
    return new MappingError(
      "duplicate_source_column",
      column,
      `duplicate source column: ${column}`,
    );
  }

  static numericBind(parameter: string): MappingError {
    return new MappingError(
      "numeric_bind",
      parameter,
      `numeric bind placeholder is not supported: ${parameter}`,
    );
  }
}

export function extractNamedBinds(sql: string): string[] {
  const binds: string[] = [];
  const seen = new Set<string>();
  let index = 0;

  while (index < sql.length) {
    const character = sql[index];

    if (character === "'" || character === '"') {
      index = skipQuoted(sql, index);
    } else if (character === "-" && sql[index + 1] === "-") {
      index = skipLineComment(sql, index + 2);
    } else if (character === "/" && sql[index + 1] === "*") {
      index = skipBlockComment(sql, index + 2);
    } else if (character === ":") {
      const start = index + 1;
      const next = sql[start];

      if (next !== undefined && isBindStart(next)) {
        let end = start + 1;
        while (end < sql.length && isBindContinue(sql[end])) {
          end += 1;
        }

        const bind = sql.slice(start, end);
        const normalized = asciiUpper(bind);
        if (!seen.has(normalized)) {
          seen.add(normalized);
          binds.push(bind);
        }
        index = end;
      } else if (next !== undefined && isAsciiDigit(next)) {
        let end = start + 1;
        while (end < sql.length && isAsciiDigit(sql[end])) {
          end += 1;
        }
        throw MappingError.numericBind(sql.slice(start, end));
      } else {
        index += 1;
      }
    } else {
      index += 1;
    }
  }

  return binds;
}

export function mapRow(row: SourceRow, bindNames: readonly string[]): NamedRow {
  const normalized = new Map<string, DomainValue>();

  for (const [column, value] of Object.entries(row)) {
    const normalizedColumn = asciiUpper(column);
    if (normalized.has(normalizedColumn)) {
      throw MappingError.duplicateSourceColumn(normalizedColumn);
    }
    normalized.set(normalizedColumn, value);
  }

  const mapped = Object.create(null) as NamedRow;
  for (const bind of bindNames) {
    const normalizedBind = asciiUpper(bind);
    if (!normalized.has(normalizedBind)) {
      throw MappingError.missingSourceColumn(bind);
    }
    mapped[bind] = normalized.get(normalizedBind) as DomainValue;
  }
  return mapped;
}

function isBindStart(character: string): boolean {
  return isAsciiLetter(character) || character === "_";
}

function isBindContinue(character: string): boolean {
  return isBindStart(character)
    || isAsciiDigit(character)
    || character === "$"
    || character === "#";
}

function isAsciiLetter(character: string): boolean {
  const code = character.charCodeAt(0);
  return (code >= 65 && code <= 90) || (code >= 97 && code <= 122);
}

function isAsciiDigit(character: string): boolean {
  const code = character.charCodeAt(0);
  return code >= 48 && code <= 57;
}

function asciiUpper(value: string): string {
  return value.replace(/[a-z]/g, (character) => character.toUpperCase());
}

function skipQuoted(sql: string, start: number): number {
  const quote = sql[start];
  let index = start + 1;

  while (index < sql.length) {
    if (sql[index] === quote) {
      if (sql[index + 1] === quote) {
        index += 2;
      } else {
        return index + 1;
      }
    } else {
      index += 1;
    }
  }

  return index;
}

function skipLineComment(sql: string, start: number): number {
  let index = start;
  while (index < sql.length && sql[index] !== "\n") {
    index += 1;
  }
  return index;
}

function skipBlockComment(sql: string, start: number): number {
  let index = start;
  while (index < sql.length) {
    if (sql[index] === "*" && sql[index + 1] === "/") {
      return index + 2;
    }
    index += 1;
  }
  return index;
}
